### ISSUE TITLE
Average Pooling layer support missing from LRP

### ISSUE THREAD

**Issue by danielbraun89** [albermax/innvestigate#94]
In order to support max-pooling layers the LRP analyzer (relevance_analyzer.py) replaces them with average pooling layers.

However, there is no support for the average pooling layer themselves. 

Is there a special reason for this odd behavior? 
When I have monkey patched the code in order to use averagePoolingLayer as a replacement for itself it works with no problem.

So the question is was there any inherent constraint in the way that LRP method work that prevents average pooling layers from being supported?

**Comment by sebastian-lapuschkin** [comment]
Hi @danielbraun89, 

AveragePooling is right now supported via `innvestigate.analyzer.relevance_based.LRP.AveragePoolingReverseLayer` in line 446, with the conditional reverse hook being defined in line 483 of the same file.

MaxPooling is covered via the `_default_reverse_mapping` function in line 492, which will be called if none of the conditions of `...LRP._conditional_mappings` apply.

Does the intended behaviour not work for your model?
Can you provide us with some mor information?

Cheers,

**Comment by danielbraun89** [comment]
I cant find LRP under innvestigate.analyzer.relevance_based,  ( looking at https://github.com/albermax/innvestigate/tree/master/innvestigate/analyzer/relevance_based ,  am I at the right place?)

**Comment by sebastian-lapuschkin** [comment]
Have a look at [https://github.com/albermax/innvestigate/blob/master/innvestigate/analyzer/relevance_based/relevance_analyzer.py](https://github.com/albermax/innvestigate/blob/master/innvestigate/analyzer/relevance_based/relevance_analyzer.py), line 231

**Comment by danielbraun89** [comment]
In further examination I suspect the problem is in the list 'BASELINELRPZ_LAYERS'
(https://github.com/albermax/innvestigate/blob/master/innvestigate/analyzer/relevance_based/relevance_analyzer.py line 97)
This list is being checked at line 166 and throws an exception whenever I use any net that contains an averagePoolingLayer as it does not appear on the list.

Is this the intentional behavior? 
If not would a net with an average pooling layer would run an LRP if I just add keras.layers.pooling.AveragePooling1/2/3D to the list?

**Comment by sebastian-lapuschkin** [comment]
The BaselineLRPZ class you are using implements Gradient*Input. This class will vanish or properly refactored as part of Issue #82
I suggest you use (or create) a subclass of LRP (as defined in line 231) which does handle AveragePooling properly (see line 483, the used function  [`kchecks.is_average_pooling`](https://github.com/albermax/innvestigate/blob/master/innvestigate/utils/keras/checks.py) (definition in line 354 in linked file).

A good starting point is to 
`lrp = innvestigate.create_analyzer('lrp.sequential_preset*')`
where you should fill in * from the list `analyzers` in [https://github.com/albermax/innvestigate/blob/master/innvestigate/analyzer/__init__.py](https://github.com/albermax/innvestigate/blob/master/innvestigate/analyzer/__init__.py)

**Comment by danielbraun89** [comment]
Understood, thanks

**Comment by sebastian-lapuschkin** [comment]
You are welcome
